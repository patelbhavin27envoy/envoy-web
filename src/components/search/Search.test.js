
require('fetch-reply-with');

import React from 'react';
import renderer from 'react-test-renderer';

import { shallow, mount } from "enzyme";
import { MemoryRouter } from 'react-router-dom';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

import Search from './Search'

describe("Search Component :: ", () => {

    it("should render without throwing an error", () => {
        expect(
            renderer.create(
                <MemoryRouter>
                    <Search/>
                </MemoryRouter>
            )
        ).not.toBe(null);
    });


    it("should render search inputs UI", () => {

        // Act
        let component = renderer.create(
            <MemoryRouter>
                <Search/>
            </MemoryRouter>
        ).toJSON();
        // snapshot
        expect(component).toMatchSnapshot();


        // Act
        component = mount(
            <MemoryRouter>
                <Search/>
            </MemoryRouter>
        );

        // Assert
        expect((component).prop('state')).not.toEqual(null);
        expect(component.find('form')).not.toEqual(null);
        expect(component.find('.py3.block')).not.toEqual(null);
        expect(component.find('.search-item-m').text()).toEqual('');
        expect(component.find('.new-visitor').text()).toContain('New visitor');
    });


    it("should change state with value change in input fields", () => {


        // define the response
        const mockSuccessResponse = {status : true, visitor : []};
        // fetch.mockResponseOnce(JSON.stringify(mockSuccessResponse));

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/visitors';
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4


        // Act
        let component = mount(
            <MemoryRouter keyLength={0}>
                <Search/>
            </MemoryRouter>
        );


        ////////////
        // Assert
        ////////////

        // basic check
        expect((component).prop('state')).not.toEqual(null);
        expect(component.find('form')).not.toEqual(null);
        expect(component.find('.py3.block')).not.toEqual(null);
        expect(component.find('.new-visitor').text()).toContain('New visitor');

        // type in the text
        component.find('.search-item-m').simulate('change', {target: {name: 'searchItem', value: 'john'}});
        expect(global.fetch).toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith(endPoint);

        component.setState({inputError: true});

    });


    it("should change state with value from API error", () => {


        // define the response
        const mockSuccessResponse = null;
        // fetch.mockResponseOnce(JSON.stringify(mockSuccessResponse));

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/visitors';
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4


        // Act
        let component = mount(
            <MemoryRouter keyLength={0}>
                <Search/>
            </MemoryRouter>
        );

        ////////////
        // Assert
        ////////////

        // basic check
        expect((component).prop('state')).not.toEqual(null);
        expect(component.find('form')).not.toEqual(null);
        expect(component.find('.py3.block')).not.toEqual(null);
        expect(component.find('.new-visitor').text()).toContain('New visitor');

        // type in the text
        component.find('.search-item-m').simulate('change', {target: {name: 'searchItem', value: 'john'}});
        expect(global.fetch).toHaveBeenCalledTimes(2);
        expect(global.fetch).toHaveBeenCalledWith(endPoint);

        component.find('.new-visitor').simulate('click');

    });


    it("should search items based on search", () => {

        // define the response
        const mockSuccessResponse = null;
        // fetch.mockResponseOnce(JSON.stringify(mockSuccessResponse));

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/visitors?name=JohnDelete';
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4


        // Act
        let component = mount(
            <MemoryRouter keyLength={0}>
            <Search/>
            </MemoryRouter>
    );

        ////////////
        // Assert
        ////////////

        // basic check
        expect((component).prop('state')).not.toEqual(null);
        expect(component.find('form')).not.toEqual(null);
        expect(component.find('.py3.block')).not.toEqual(null);
        expect(component.find('.search-item-s').text()).toEqual('');
        expect(component.find('.new-visitor').text()).toContain('New visitor');

        // type in the text
        component.find('.search-item-m').simulate('change', {target: {name: 'searchItem', value: 'JohnDelete'}});
        expect(global.fetch).toHaveBeenCalledTimes(3);

        component.find('.search').first().simulate('click');
        expect(global.fetch).toHaveBeenCalledTimes(4);
        expect(global.fetch).toHaveBeenCalledWith(endPoint);

    });



});

