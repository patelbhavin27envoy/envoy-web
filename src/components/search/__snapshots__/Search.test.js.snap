// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component ::  should render search inputs UI 1`] = `
<div
  className="field"
  role="search"
>
  <div
    className="space-wrapper"
  />
  <button
    className="btn  btn-brand float-right ml-2 new-visitor"
    onClick={[Function]}
  >
    <i
      className="fas fa-user"
    />
      New visitor
  </button>
  <button
    className="btn  btn-brand float-right ml-2 search non-mobile"
    onClick={[Function]}
    type="submit"
  >
    <i
      className="fas fa-search"
    />
  </button>
  <input
    className="search-item-m p-2 text-sm border float-right max-w-xs w-full non-mobile"
    name="searchItem"
    onChange={[Function]}
    placeholder="Search"
    type="text"
    value=""
  />
  <img
    alt="Envoy Logo"
    className="py3 block"
    src="https://dashboard.envoy.com/assets/images/logo-small-red-ba0cf4a025dd5296cf6e002e28ad38be.svg"
    width="31"
  />
  <div
    className="space-wrapper"
  />
  <div
    className="mobile"
  >
    <input
      className="search-item-s p-2 text-sm border max-w-xs"
      name="searchItem"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      className="btn  btn-brand  p-2 search "
      onClick={[Function]}
      type="submit"
    >
      <i
        className="fas fa-search"
      />
    </button>
  </div>
</div>
`;
